tests/std/tests/P2502R2_generator/test.cpp

```diff
@@ -0,0 +1,429 @@
+void test_one(generator<R, V, A> g0, invocable<generator<R, V, A>> auto adaptor, ranges::input_range auto&& expected)
+    requires ranges::input_range<decltype(adaptor(move(g0)))>
+{
+    static_assert(same_as<generator<R, V, A>, typename Traits::generator>);
+    static_assert(static_checks<Traits, ValueType, ReferenceType, RvalueReferenceType>());
+
+    auto g1 = move(g0);
+    auto i  = ranges::cbegin(expected);
+    for (auto&& x : adaptor(move(g1))) {
+        assert(i != ranges::cend(expected));
+        assert(*i == x);
+        ++i;
+        // Verify iterator stays valid after move assignment
+        ranges::swap(g0, g1);
+    }
+    assert(i == ranges::cend(expected));
+}
+
+template <class Traits, class ValueType, class ReferenceType, class RvalueReferenceType, class R, class V, class A,
+    ranges::input_range Ex>
+void test_one(generator<R, V, A> g0, Ex&& expected) {
+    return test_one<Traits, ValueType, ReferenceType, RvalueReferenceType>(move(g0), identity{}, forward<Ex>(expected));
+}
+
+// Some simple end-to-end tests, mostly from the Working Draft or P2502R2
+generator<int> ints(int start = 0) {
+    while (true) {
+        co_yield start++;
+    }
+}
+
+template <ranges::input_range Rng1, ranges::input_range Rng2>
+generator<tuple<ranges::range_reference_t<Rng1>, ranges::range_reference_t<Rng2>>,
+    tuple<ranges::range_value_t<Rng1>, ranges::range_value_t<Rng2>>>
+    co_zip(Rng1 r1, Rng2 r2) {
+    auto it1        = ranges::begin(r1);
+    auto it2        = ranges::begin(r2);
+    const auto end1 = ranges::end(r1);
+    const auto end2 = ranges::end(r2);
+    for (; it1 != end1 && it2 != end2; ++it1, ++it2) {
+        co_yield {*it1, *it2};
+    }
+}
+
+void zip_example() {
+    using V = tuple<int, int>;
+    using R = tuple<int&, int&>;
+
+    auto g0 = co_zip(array{1, 2, 3}, vector{10, 20, 30, 40, 50});
+    test_one<gen_traits<R, V>, V, R, R>(move(g0), array{tuple{1, 10}, tuple{2, 20}, tuple{3, 30}});
+
+    g0 = co_zip(array{3, 2, 1}, vector{10, 20, 30, 40, 50});
+    test_one<gen_traits<R, V>, V, R, R>(move(g0), array{tuple{3, 10}, tuple{2, 20}, tuple{1, 30}});
+}
+
+template <class Reference = const int&>
+generator<Reference, int> co_upto(const int hi) {
+    assert(hi >= 0);
+    for (int i = 0; i < hi; ++i) {
+        co_yield i;
+    }
+}
+
+void test_weird_reference_types() {
+    constexpr int n = 32;
+    { // Test mutable lvalue reference type
+        auto r   = co_upto<int&>(n);
+        auto pos = r.begin();
+        for (int i = 0; i < n / 2; ++i, ++*pos, ++pos) {
+            assert(pos != r.end());
+            assert(*pos == 2 * i);
+        }
+        assert(pos == r.end());
+    }
+
+#if !(defined(__clang__) && defined(_M_IX86)) // TRANSITION, LLVM-56507
+    { // Test with mutable xvalue reference type
+        auto woof = [](size_t size, size_t count) -> generator<vector<int>&&> {
+            random_device rd{};
+            uniform_int_distribution dist{0, 99};

```

Comment:
This is producing truly random values, but not logging them. (We have a system for randomized tests where we truly-random seed a Mersenne Twister and log the seed, so we can reproduce failures if necessary.) But then I observe that this test doesn't appear to actually care about any of the values being generated into the vector - it looks like they could all be constant. The use of `dist(rd)` therefore appears to be a distraction.

Could we just fill the vector with a constant? (Or if variation is needed for some reason I'm missing, `static_cast<int>(count)`?) Additionally, it looks like this might be the only use of `<random>` in the test, if so the header should be dropped.

Reply:
There is just _way_ too much extraneous cruft here. Let's distill this down to just what it's actually testing.